OSASINFRA-3311: openstack: CAPO provider deployment

EmilienM opened the pull request:
**What this PR does / why we need it**:

- **openstack: add CAPO deployment and OpenStackCluster object**
- **Run `make update`**

This is the third PR to support OpenStack as dev-preview in 4.17.
This PR focuses on the CAPI provider. There are 2 commits:
* One that reviewers should review `1004 insertions(+), 146 deletions(-)`
* One that was generated by `make update`

EmilienM commented:
/cc csrwng bryan-cox

EmilienM commented:
/test unit
sounds like a flake in a unrelated test.

EmilienM commented:
/test e2e-aks

bryan-cox commented on api/hypershift/v1beta1/hostedcluster_types.go:
```diff
@@ -1832,8 +1836,217 @@ type AzurePlatformSpec struct {
+type SubnetSpec struct {
+	// DNSNameservers holds a list of DNS server addresses that will be provided when creating
+	// the subnet. These addresses need to have the same IP version as CIDR.
+	DNSNameservers []string `json:"dnsNameservers,omitempty"`
```
Can we specifically say whether or not this is required or optional please?

bryan-cox commented on api/hypershift/v1beta1/hostedcluster_types.go:
```diff
@@ -1832,8 +1836,217 @@ type AzurePlatformSpec struct {
+	// AllocationPools is an array of AllocationPool objects that will be applied to OpenStack Subnet being created.
+	// If set, OpenStack will only allocate these IPs for Machines. It will still be possible to create ports from
+	// outside of these ranges manually.
+	AllocationPools []AllocationPool `json:"allocationPools,omitempty"`
```
Can we specifically say whether or not this is required or optional please?

bryan-cox commented on api/hypershift/v1beta1/hostedcluster_types.go:
```diff
@@ -1832,8 +1836,217 @@ type AzurePlatformSpec struct {
+
+type AllocationPool struct {
+	// Start represents the start of the AllocationPool, that is the lowest IP of the pool.
+	// +required
```
Does this need `// +kubebuilder:validation:Required`?

bryan-cox commented on api/hypershift/v1beta1/hostedcluster_types.go:
```diff
@@ -1832,8 +1836,217 @@ type AzurePlatformSpec struct {
+	Start string `json:"start"`
+
+	// End represents the end of the AlloctionPool, that is the highest IP of the pool.
+	// +required
```
Does this need `// +kubebuilder:validation:Required`?

bryan-cox commented on api/hypershift/v1beta1/hostedcluster_types.go:
```diff
@@ -1832,8 +1836,217 @@ type AzurePlatformSpec struct {
+	ID *string `json:"id,omitempty"`
+
+	// Filter specifies a filter to select an OpenStack router. If provided, cannot be empty.
+	Filter *RouterFilter `json:"filter,omitempty"`
```
Can we specifically say whether or not this is required or optional please?

bryan-cox commented on api/hypershift/v1beta1/hostedcluster_types.go:
```diff
@@ -1832,8 +1836,217 @@ type AzurePlatformSpec struct {
+
+// RouterFilter specifies a query to select an OpenStack router. At least one property must be set.
+// +kubebuilder:validation:MinProperties:=1
+type RouterFilter struct {
```
For each property, can we specifically say whether or not this is required or optional please?

bryan-cox commented on api/hypershift/v1beta1/hostedcluster_types.go:
```diff
@@ -1832,8 +1836,217 @@ type AzurePlatformSpec struct {
+
+// NetworkFilter specifies a query to select an OpenStack network. At least one property must be set.
+// +kubebuilder:validation:MinProperties:=1
+type NetworkFilter struct {
```
For each property, can we specifically say whether or not this is required or optional please?

bryan-cox commented on api/hypershift/v1beta1/hostedcluster_types.go:
```diff
@@ -1832,8 +1836,217 @@ type AzurePlatformSpec struct {
+
+// SubnetFilter specifies a filter to select a subnet. At least one parameter must be specified.
+// +kubebuilder:validation:MinProperties:=1
+type SubnetFilter struct {
```
For each property, can we specifically say whether or not this is required or optional please?

bryan-cox reviewed (commented):
Mostly API comments. We will be going through an effort to mark all API fields as required or optional explicitly; trying to save reworking through these in the future 😄

EmilienM commented:
> Mostly API comments. We will be going through an effort to mark all API fields as required or optional explicitly; trying to save reworking through these in the future 😄

Thanks, great feedback. We'll keep that in mind for our future API changes.
I think I addressed the comments.

bryan-cox commented:
/approve

EmilienM commented:
/retest e2e-azure

EmilienM commented:
/test e2e-azure

csrwng commented:
/lgtm
